From b5afaaad135ca28781a7510cc6dbe353f65d50f0 Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Tue, 19 Aug 2014 19:30:43 -0400
Subject: [scsi] qla4xxx: Add support for ISCSI_PARAM_LOCAL_IPADDR sysfs attr

Message-id: <1408476658-5056-6-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88341
O-Subject: [PATCH RHEL 7.1 e-stor 05/20] qla4xxx: Add support for ISCSI_PARAM_LOCAL_IPADDR sysfs attr
Bugzilla: 1089349
RH-Acked-by: Rob Evers <revers@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Adheer Chandravanshi <adheer.chandravanshi@qlogic.com>

scsi commit id fb734ee3eff40c1374e23be6eaf666681e6137ac

Signed-off-by: Adheer Chandravanshi <adheer.chandravanshi@qlogic.com>
Signed-off-by: Vikas Chaudhary <vikas.chaudhary@qlogic.com>
Reviewed-by: Mike Christie <michaelc@cs.wisc.edu>
Signed-off-by: James Bottomley <JBottomley@Parallels.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/qla4xxx/ql4_os.c b/drivers/scsi/qla4xxx/ql4_os.c
index 31ff616..cc8fbb3 100644
--- a/drivers/scsi/qla4xxx/ql4_os.c
+++ b/drivers/scsi/qla4xxx/ql4_os.c
@@ -421,6 +421,7 @@ static umode_t qla4_attr_is_visible(int param_type, int param)
   case ISCSI_PARAM_EXP_STATSN:
   case ISCSI_PARAM_DISCOVERY_PARENT_IDX:
   case ISCSI_PARAM_DISCOVERY_PARENT_TYPE:
+  case ISCSI_PARAM_LOCAL_IPADDR:
    return S_IRUGO;
   default:
    return 0;
@@ -3714,6 +3715,7 @@ static void qla4xxx_copy_to_sess_conn_params(struct iscsi_conn *conn,
  unsigned long options = 0;
  uint16_t ddb_link;
  uint16_t disc_parent;
+ char ip_addr[DDB_IPADDR_LEN];
 
  options = le16_to_cpu(fw_ddb_entry->options);
  conn->is_fw_assigned_ipv6 = test_bit(OPT_IS_FW_ASSIGNED_IPV6, &options);
@@ -3795,6 +3797,14 @@ static void qla4xxx_copy_to_sess_conn_params(struct iscsi_conn *conn,
 
  iscsi_set_param(conn->cls_conn, ISCSI_PARAM_TARGET_ALIAS,
    (char *)fw_ddb_entry->iscsi_alias, 0);
+
+ options = le16_to_cpu(fw_ddb_entry->options);
+ if (options & DDB_OPT_IPV6_DEVICE) {
+  memset(ip_addr, 0, sizeof(ip_addr));
+  sprintf(ip_addr, "%pI6", fw_ddb_entry->link_local_ipv6_addr);
+  iscsi_set_param(conn->cls_conn, ISCSI_PARAM_LOCAL_IPADDR,
+    (char *)ip_addr, 0);
+ }
 }
 
 static void qla4xxx_copy_fwddb_param(struct scsi_qla_host *ha,
-- 
1.7.1